docs/source/rfc/isa_restructuring.org

```diff
@@ -0,0 +1,515 @@
+manifests as having too many instructions: the current exhaustive list
+o finstructions is the following enumeration:
+
+#+begin_src rust
+  pub enum InstructionBox {
+          Register(RegisterBox),
+          Unregister(UnregisterBox),
+          Mint(MintBox),
+          Burn(BurnBox),
+          Transfer(TransferBox),
+          If(Box<Conditional>),
+          Pair(Box<Pair>),
+          Sequence(SequenceBox),
+          SetKeyValue(SetKeyValueBox),
+          RemoveKeyValue(RemoveKeyValueBox),
+          Grant(GrantBox),
+          Revoke(RevokeBox),
+          ExecuteTrigger(ExecuteTriggerBox),
+          SetParameter(SetParameterBox),
+          NewParameter(NewParameterBox),
+          Upgrade(UpgradeBox),
+          Log(LogBox),
+          Fail(FailBox),
+   }
+#+end_src
+
+Consider the redundancy: the same conceptual notion of =Add=, is
+represented in =Register= as in add something to a collection, but
+also in =Mint= which means add to a collection of something,
+=SetKeyValue=, which is to add a =Value= to a certain associative
+container, =Grant= to add a certain permission or =Role= (which is a
+concept specific to private blockchains), to a user.
+
+The redundancy is exacerbated because the inverse processes for each
+of these operations are also represented. The main reason for such an
+implementation is to allow for certain processes to exist without
+their inverses, and for that to be a compile-time rather than run-time
+error. In addition to leading to terrible discoverability in languages
+other than Rust, this leads the user to believe that if an instruction
+compiles, it will run. There are multiple ways of achieving the same
+result that do not force the user to have separated (sometimes by
+multiple hundreds of lines of code) implementations of the forward and
+inverse operations.
+
+Furthermore, because =Expressions= are used mainly as arguments, and
+not as general purpose-combination devices, the same devices need to
+be replicated in the instruction set as =InstructionBox= variants.
+
+We must now talk about the instruction application procedure. The
+instructions can have multiple side-effects. As such, because the
+instructions represent atomic[fn:1] =WorldStateView= modifications, if a
+particular instruction in a transaction fails to apply, all
+instructions effects must be reversed. But in the given model, such a
+concept might not be possible to define, and as such, the only way to
+enforce atomicity, given that any transaction can affect potentially
+an unconstrained number of objects, the result is that the largest
+object in the blockchain must be cloned before each application. There
+are relatively efficient ways of mitigating this with persistent data
+structures but all of them pale in comparison with the efficiency of
+having a simpler instruction set in which a tiny addition to a number
+is reverted by changing that tiny number back. This is not impossible
```

Comment:
I think this claim is not that obvious.
With current default configuration we can have up to 4096 instructions per transaction if we will adopt RISC isi we can expect this number to increase. In such case reverting N instructions might be more costly then using something like MVCC which is commonly used for database management systems, which also can provide us WSV view at particular height, which might be very useful for heavy quires to provide snapshot isolation and providing finalized wsv to resolve soft-forks.

Reply:
Isn't something what @SamHSmith wanted to implement using versions?

Reply:
Also as I understand the idea is to have complex instructions implemented as a DSL inside Executor. That way we could also implement reverting instructions inside Executor.

I would expect that we could decrease the amount of complex instructions used by user in transaction but increase the amount of native instructions used by Executor to express complex instructions.

Reply:
> Isn't something what @SamHSmith wanted to implement using versions?

I was about to say that. 

MVCC doesn't care if the instruction set is RISC (and only one of the proposals entails using a RISC ISA). 

It is true that we can have up to 4096 instructions per transaction. It is also true that under this proposal, frequently used combinations are to be encoded as expressions and not as bare instructions, so the newly proposed blockchain is likely to be smaller, not larger. 

In variant 2, I propose having RISC, but of a very particular kind. It comprises essentially the fragment of the final state, so the collection of instructions can be viewed as a `WorldStateView`. What this means is that @SamHSmith will get to implement something very similar to working sets, but with a generalised application. 

> I would expect that we could decrease the amount of complex instructions used by user in transaction but increase the amount of native instructions used by Executor to express complex instructions.

Indeed that's the intention. The idea is that RISC is a very general flexible language that is to be generated by smartcontracts, and which can be sent from any compliant client to any compliant host (in which case the executor simply validates the transaction). 

Any other complex operation which is likely to be re-used, is an expression with implementation-defined representation. So if Iroha is running a scientific lab, or a set of smart meters where none of our domain-specific financial instructions are useful, most of the blockchain contains `ExpressionBox` variants, which make sense for this application.